### PR TITLE
Create rootfs for tests

### LINKED ISSUE #265: Create a suitable bundle for functional tests
RIght now we need to create a test bundle by hand for functional tests and provide its path to the configure script. We should create it as part of "make check" instead so we can:

  - automate its creation
  - have everyone testing with the same environment
  - control what goes in there for special testing (we could have binaries in that bundle we need to run to provide a testable output), eg. test signal forwarding to a process inside the VM.

### PR DESCRIPTION
**Pull request by jodh-intel**
This is quite a tricky problem. The solution I've come up with is a slight abuse of `configure`: it would be better to have the bundle creation and deletion handled in `Makefile.am` but that would require the makefile to handle the generation of `tests/functional/common.bash.in`. That's possible, but gets rather complicated given the number of tests and options required to be checked. Note also that neither the tests themselves nor `tests/functional/common.bash.in` can handle the bundle creation/deletion since that would result in the bundle being created 'n' times, not just once.

Fixes #265.

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @gorozco1 - I've now fixed the issue caught by Travis, so removing the do-not-merge label...

File: Makefile.am [hunk 1]
```diff
@@ -206,7 +207,12 @@ endif
 if FUNCTIONAL_TESTS
 CHECK_DEPS += functional-tests
 functional-tests: cc-oci-runtime
+	$(AM_V_GEN)test -n "$(BUNDLE_TEST_PATH)" && \
+		echo "Using bundle '$(BUNDLE_TEST_PATH)'" || true
 	@$(BATS_PATH) $(srcdir)/tests/functional/
+	$(AM_V_GEN)test -e "$(BUNDLE_TEST_PATH)" -a "$(REMOVE_BUNDLE_ON_SUCCESS)" = "yes" \
```
Comment (dlespiau) [inline Makefile.am]: Random comment (nothing to do here): I think you're over doing it :) removing the bundle on "make clean" would have been enough?

File: configure.ac [hunk 2]
```diff
@@ -192,34 +192,59 @@ AC_SUBST([QEMU_PATH], [$ac_cv_path_QEMU])
...
         BUNDLE_TEST_PATH=$(cd "$with_tests_bundle_path"; pwd -P)
         AC_SUBST([BUNDLE_TEST_PATH])
-    ],
-    [
-        AC_SUBST([BUNDLE_TEST_PATH], [$DEFAULT_BUNDLE_TEST_DIR])
     ]
 )
 
-DEFAULT_ROOTFS_PATH=$BUNDLE_TEST_PATH/rootfs
 AC_ARG_WITH([tests-rootfs-path],
```
Comment (dlespiau) [inline configure.ac]: could we simplify the configuration here? (just an innocent question, shouldn't be part of this PR). is there an interesting case where we want ROOTFS_PATH != BUNDLE_PATH/rootfs?

File: configure.ac [hunk 3]
```diff
@@ -192,34 +192,59 @@ AC_SUBST([QEMU_PATH], [$ac_cv_path_QEMU])
...
+AC_CHECK_PROG([DOCKER_AVAILABLE], [docker], [yes], [no])
+
+AS_IF([test x"$with_tests_rootfs_path" = x"" dnl
+       -a x"$with_tests_bundle_path" = x"" dnl
+       -a x"$disable_tests" = xno dnl
+       -a x"$disable_functional_tests" = xno dnl
+       -a x"$auto_bundle_creation" = xyes ], [
+       AS_IF([test x"$DOCKER_AVAILABLE" = x"yes" -a x"$MKTEMP_AVAILABLE" = x"yes"],
+           [
+               BUNDLE_TEST_PATH=$([sh data/make-bundle-dir.sh])
```
Comment (dlespiau) [inline configure.ac]: That does look somewhat ugly. How about generating the bundle in a sub-directory of tests/?

An alternative could be generating the test bundle with its own Makefile rule: mid-term we could build binaries to include in the test bundle (eg. test signal handling with a binary printing the name of received signal) and track the dependencies to rebuild the bundle (which would include the Makefile itself).

Comment (dlespiau) [inline configure.ac]: oh, and that was in the PR message... oh well.

Comment (dlespiau) [inline configure.ac]: Actually, if the problem is only the generation of tests/functional/common.bash, it's easy to do it from the Makefile with a sed, all the variables I see at the top of that file are exposed to the Makefile.

**Comment by dlespiau** [comment]
A few comments I'd love your thoughts about. Using a temp dir in the configure lasting for the lifetime of said configure (which can span machine reboots :/) is a bit meh.

File: configure.ac (hunk 2 above)
Comment (jodh-intel) [inline configure.ac]: Not really. I think you're right - I'll remove the rootfs option..

File: Makefile.am (hunk 1 above)
Comment (jodh-intel) [inline Makefile.am]: Good point - I'll do this...

File: configure.ac (hunk 3 above)
Comment (jodh-intel) [inline configure.ac]: Right - having the configure stage generate the rootfs is pretty yucky as that operation should be part of the build.

After basically telling myself how I should have done it the first time, I'll take my own advice I think! :-)

**Comment by jodh-intel** [comment]
Relabelled to allow for a rework.

**Comment by jodh-intel** [comment]
Rework now complete, so this is ready for re-review.

File: tests/functional/common.bash.in [hunk 4]
```diff
@@ -34,7 +34,7 @@ CONFIG="${TEST_DATA_DIR}/config-minimal-cc-oci.json"
 HYPERVISOR_ARGS="${BATS_TEST_DIRNAME}/../../data/hypervisor.args"
 
 BUNDLE_DIR="@BUNDLE_TEST_PATH@"
-ROOTFS_DIR="@BUNDLE_TEST_PATH@"
+[ -n "$BUNDLE_DIR" ] && ROOTFS_DIR="$BUNDLE_DIR/rootfs"
```
Comment (jodh-intel) [inline tests/functional/common.bash.in]: In that case, `ROOTFS_DIR` would also be empty. That's what we want as `ceck_ccontainers()` checks both variables and if they don't exist as directories (they won't if they are unset), the tests that needs bundles/rootfs's are skipped.

Comment (gorozco1) [inline tests/functional/common.bash.in]: oh great thanks for clarify :)

File: data/make-bundle-dir.sh [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+# contain a sub-directory called "rootfs" (which is the canonical path to
+# the rootfs specified by an OCI bundle).
+#
+# See the following for further details:
+#
+# - https://github.com/opencontainers/runtime-spec/blob/master/bundle.md
+
+# Name of a small docker image that create_rootfs() uses
+# to make a chroot directory.
+docker_image="busybox"
```
Comment (devimc) [inline data/make-bundle-dir.sh]: Hi @jodh-intel 
why do not drink our own poison? https://hub.docker.com/_/clearlinux/    :-)

Comment (gorozco1) [inline data/make-bundle-dir.sh]: @devimc I think the intention here is to make this test as fast as possible and busybox image is small enough

Comment (jodh-intel) [inline data/make-bundle-dir.sh]: That's a reasonable question :-) I simply chose busybox as it's really small (< 1 MB) and so fast to download (the clearlinux docker image is ~30 MB).

We could add a `--with-auto-bundle-creation-docker-image=` if folk want to vary the image I guess. But maybe the simplest fix is to make `data/make-bundle-dir.sh` accept an optional docker image name too. If not specified, it would default to `busybox`. Then users could:

```
# create a rootfs dir from a docker clearlinux image
$ data/make-bundle-dir.sh /some/where clearlinux

$ ./configure --with-tests-bundle-path=/some/where
```

Comment (devimc) [inline data/make-bundle-dir.sh]: :+1:

**Comment by devimc** [comment]
:+1:

**Comment by devimc** [comment]
:+1:

**Comment by devimc** [comment]
sorry I closed this PR, could someone merge it ?

**Comment by dlespiau** [comment]
(the lgtm needs to be in a comment, not the github's approved ones)

lgtm

[Approved with PullApprove](https://pullapprove.com/01org/cc-oci-runtime/pull-request/272/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=01org/cc-oci-runtime)
